test_rclcpp/test/test_subscription.cpp

```diff
@@ -371,6 +371,110 @@ TEST(CLASSNAME(test_subscription, RMW_IMPLEMENTATION), subscription_shared_ptr_c
   ASSERT_EQ(1, counter);
 }
 
+// Shortened version of the test for subscribing after spinning has started.
+TEST(CLASSNAME(test_subscription, RMW_IMPLEMENTATION), spin_before_subscription) {
+  auto node = rclcpp::Node::make_shared("spin_before_subscription");
+
+  //pub_qos_profile.durability = RMW_QOS_POLICY_TRANSIENT_LOCAL_DURABILITY;
+  auto publisher = node->create_publisher<test_rclcpp::msg::UInt32>(
+    "spin_before_subscription", rmw_qos_profile_default);
+
+  uint32_t counter = 0;
+  auto callback =
+    [&counter](test_rclcpp::msg::UInt32::ConstSharedPtr msg) -> void
+    {
+      ++counter;
+      printf("  callback() %4u with message data %u\n", counter, msg->data);
+      ASSERT_EQ(counter, msg->data);
+    };
+
+  auto msg = std::make_shared<test_rclcpp::msg::UInt32>();
+  msg->data = 0;
+  rclcpp::executors::SingleThreadedExecutor executor;
+  executor.add_node(node);
+
+  executor.spin_some();
+
+  auto subscriber = node->create_subscription<test_rclcpp::msg::UInt32>(
+    "spin_before_subscription", callback, rmw_qos_profile_default);
+
+  // start condition
+  ASSERT_EQ(0, counter);
+
+  size_t i = 0;
+  while (node->count_subscribers("spin_before_subscription") == 0 && ++i < max_loops) {
```

Comment:
I've redone this logic a little bit by putting it into a function and using a more readable counting method for the busy wait. I'm also reusing the function throughout the tests in this file to wait for a subscription to register for a certain topic, rather than waiting a fixed amount of time.